fix compile DTensor.from_local in trace_rule_look up

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119659

There's a bug when converting from TorchVariable to trace rule look ups,
in some corner cases the DTensor.from_local calls not matching the trace
name rule look up, resulting in a None look up, and falling back to the
UserFunctionVariable, which makes the tracing silent wrong by tracing
into the DTensor.from_local function. Not exactly sure yet why the look
up failed

This PR fixes the DTensor.from_local tracing to make sure in everycase
we should hit the InGraphFunctionVariable

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

wanchaol committed:
fix compile DTensor.from_local in trace_rule_look up

There's a bug when converting from TorchVariable to trace rule look ups,
in some corner cases the DTensor.from_local calls not matching the torch
name rule look up, resulting in a None look up, and falling back to the
UserFunctionVariable, which makes the tracing silent wrong by tracing
into the DTensor.from_local function. Not exactly sure yet why the look
up failed

This PR fixes the DTensor.from_local tracing to make sure in everycase
we should hit the InGraphFunctionVariable

[ghstack-poisoned]

yifuwang commented on test/distributed/_tensor/test_dtensor_compile.py:
```diff
@@ -302,9 +310,12 @@ def forward(self, input):
         inp = torch.rand(20, 16).to(self.device_type)
         out = compiled_model(inp)
+        out.sum().backward()
+        self.assertEqual(cnt.frame_count, 1)
```
Nice!

yifuwang reviewed (approved):
Thanks a lot for root causing and fixing this so quickly! The test looks awesome!

wanchaol committed:
Update on "fix compile DTensor.from_local in trace_rule_look up"


There's a bug when converting from TorchVariable to trace rule look ups,
in some corner cases the DTensor.from_local calls not matching the torch
name rule look up, resulting in a None look up, and falling back to the
UserFunctionVariable, which makes the tracing silent wrong by tracing
into the DTensor.from_local function. Not exactly sure yet why the look
up failed

This PR fixes the DTensor.from_local tracing to make sure in everycase
we should hit the InGraphFunctionVariable

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

Skylion007 commented:
This seems like a good backport candidate

wanchaol commented:
@pytorchbot merge

Skylion007 commented:
@wanchaol Is this issue present in 2.2? It might be a good candidate to be backported there: https://github.com/pytorch/pytorch/issues/119295

yanboliang commented on torch/_dynamo/trace_rules.py:
```diff
@@ -92,7 +92,7 @@
     "torch.distributed.get_rank": TorchInGraphFunctionVariable,
     "torch.distributed.get_world_size": TorchInGraphFunctionVariable,
-    "torch.distributed._tensor.DTensor#from_local": TorchInGraphFunctionVariable,
+    "torch.distributed._tensor.api.DTensor#from_local": TorchInGraphFunctionVariable,
```
Should we put both of them here? It seems they are not identical, so probably others use the previous one, so we should put them in graph for both case?
Actually inside of the variable, we are using the previous one:
https://github.com/pytorch/pytorch/blob/34a61c527b242e13d617f853dcaa93d41b852320/torch/_dynamo/variables/distributed.py#L40-L45

It seems we have to cover both cases if they are not identical.

wanchaol replied:
IMO they should be identical objects, it's just that `torch.distributed._tensor.DTensor` is imported from `torch.distributed._tensor.api.DTensor`.  code [link](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/__init__.py#L9)

But indeed in some cases it does not hit the trace name map... which is really weird, it would be good if we could follow up what happened there.

If you want to be extra safe I can add both to the list, but I think the current change should capture the previous case as well, lmk :)

yanboliang replied:
SG!

wanchaol replied:
cool I think I'll land this as is to avoid trigger CI again, we can always add the previous one later if we hit problems.

wanchaol commented:
@pytorchbot merge
